File: Solution/src/com/solution/tree/TopView.java
package com.solution.tree;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.solution.main.BTreePrinter;
import com.solution.main.ReadInput;
import com.solution.model.Tree;

public class TopView {

	int lindex = 0;
	int rindex = 0;

	public void run() {

		Tree t = ReadInput.getTree("TopView");

		Map<Integer, List<Tree>> map = new HashMap<Integer, List<Tree>>();

		BTreePrinter.printNode(t);

		System.out.println("Without Hashing");
		topViewWithoutHashing(t);
		doTopView(t);

	}

	class QItem {
		public QItem(Tree t, int index) {
			this.t = t;
			this.index = index;
		}

		public Tree t;
		public int index;
	}

	private void doTopView(Tree tree) {

		ArrayDeque<QItem> q = new ArrayDeque<>();
		Map<Integer, Tree> map = new HashMap<Integer, Tree>();

		q.add(new QItem(tree, 0));

		while (!q.isEmpty()) {

			QItem qitem = q.poll();
			Tree t = qitem.t;
			if (!map.containsKey(qitem.index)) {
				map.put(qitem.index, qitem.t);
			}

			if (t.getLeft() != null) {
				q.add(new QItem(t.getLeft(), qitem.index - 1));
			}
			if (t.getRight() != null) {
				q.add(new QItem(t.getRight(), qitem.index + 1));
			}
		}

		System.out.println("Hashing:");
		map.forEach((k, v) -> {
			System.out.print(v.getData() + " ");
		});
	}

	private void topViewWithoutHashing(Tree t) {
		int height = getHeight(t);

		for (int i = 0; i < height; i++) {
			doprintTopView(t, 0, i);
		}

	}

	private void doprintTopView(Tree t, int index, int level) {

		if (t == null) {
			return;
		}

		if (level == 0) {
			if (index <= lindex) {
				lindex = index;
				System.out.print(t.getData() + " ");
			} else if (index >= rindex) {
				rindex = index;
				System.out.print(t.getData() + " ");
			}
		} else {
			doprintTopView(t.getLeft(), index - 1, level - 1);
			doprintTopView(t.getRight(), index + 1, level - 1);
		}

	}

	private int getHeight(Tree t) {
		if (t == null) {
			return 0;
		}
		int lheight = getHeight(t.getLeft());
		int rheight = getHeight(t.getRight());

		return Math.max(lheight, rheight) + 1;
	}
}

File: src/main/kotlin/com/cssolutions/tree/TopView.kt
package com.cssolutions.tree

import com.cssolutions.main.BTreePrinter.printNode
import com.cssolutions.main.ReadInput.getTree
import com.cssolutions.model.Tree
import java.util.*

class TopView {
    var lindex = 0
    var rindex = 0
    fun run() {
        val t = getTree("TopView")
        val map: Map<Int, List<Tree>> = HashMap()
        printNode<Comparable<*>>(t)
        println("Without Hashing")
        topViewWithoutHashing(t)
        doTopView(t)
    }

    internal inner class QItem(var t: Tree?, var index: Int)

    private fun doTopView(tree: Tree?) {
        val q = ArrayDeque<QItem>()
        val map: MutableMap<Int, Tree?> = HashMap()
        q.add(QItem(tree, 0))
        while (!q.isEmpty()) {
            val qitem = q.poll()
            val t = qitem.t
            if (!map.containsKey(qitem.index)) {
                map[qitem.index] = qitem.t
            }
            if (t!!.left != null) {
                q.add(QItem(t.left, qitem.index - 1))
            }
            if (t.right != null) {
                q.add(QItem(t.right, qitem.index + 1))
            }
        }
        println("Hashing:")
        map.forEach { (k: Int?, v: Tree?) -> print(v!!.data.toString() + " ") }
    }

    private fun topViewWithoutHashing(t: Tree?) {
        val height = getHeight(t)
        for (i in 0 until height) {
            doprintTopView(t, 0, i)
        }
    }

    private fun doprintTopView(t: Tree?, index: Int, level: Int) {
        if (t == null) {
            return
        }
        if (level == 0) {
            if (index <= lindex) {
                lindex = index
                print(t.data.toString() + " ")
            } else if (index >= rindex) {
                rindex = index
                print(t.data.toString() + " ")
            }
        } else {
            doprintTopView(t.left, index - 1, level - 1)
            doprintTopView(t.right, index + 1, level - 1)
        }
    }

    private fun getHeight(t: Tree?): Int {
        if (t == null) {
            return 0
        }
        val lheight = getHeight(t.left)
        val rheight = getHeight(t.right)
        return Math.max(lheight, rheight) + 1
    }
}